File: 1035-cousins-in-binary-tree/1035-cousins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        def BFS(root,x,y):
            q=deque([(root,None)])
            while q:
                l=len(q)
                ans=[]
                for _ in range(l):
                    r,p=q.popleft()
                    if r.val==x:
                        if p:
                            ans.append(p.val)
                    if r.val==y:
                        if p:
                            ans.append(p.val)
                    if r.left:
                        q.append((r.left,r))
                    if r.right:
                        q.append((r.right,r))
                if len(set(ans))==2:
                    return True
            return False
        return BFS(root,x,y)

File: 1035-cousins-in-binary-tree/1035-cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean BFS(TreeNode root,int x,int y){
        int l,i;
        Queue<Pair<TreeNode,TreeNode>> q=new LinkedList<>();
        q.add(new Pair<>(root,null));
        while(!q.isEmpty()){
            l=q.size();
            List<TreeNode> ans=new ArrayList<>();
            for(i=0;i<l;i++){
                Pair<TreeNode,TreeNode> current=q.poll();
                TreeNode r=current.getKey();
                TreeNode p=current.getValue();
                if(r.val==x||r.val==y){
                    if(p!=null){
                        ans.add(p);
                    }
                }
                if(r.left!=null){
                    q.add(new Pair<>(r.left,r));
                }
                if(r.right!=null){
                    q.add(new Pair<>(r.right,r));
                }
            }
            if(ans.size()==2 && ans.get(0)!=ans.get(1)){
                return true;
            }
        }
        return false;
    }
    public boolean isCousins(TreeNode root, int x, int y) {
        return BFS(root,x,y);
    }
}